design/one-pager-change-logs.md

```diff
@@ -0,0 +1,410 @@
+state during each reconciliation, but that would paint a less complete picture
+compared to observing the external system and recording the "after" state
+immediately after the operation as well. If we find during implementation and
+testing that recording the "after" state places too much burden on the system,
+we can revisit this decision.
+
+An additional information field is available in each change log entry for
+Providers to store any provider specific data they choose. Providers are not
+required to store any data at all in this field, as it is optional. As an
+example of data that could be stored here, `provider-aws` could store the
+specific endpoints it called to perform the given operation captured in the
+change log entry.
+
+### Generating Change Logs
+
+It is a goal that the change log data is generated as close to the source of
+truth as possible. All Crossplane providers, both [classic
+providers](https://github.com/crossplane/provider-template) and [Upjet generated
+providers](https://github.com/crossplane/upjet-provider-template) make use of
+the crossplane-runtime [managed
+reconciler](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go),
+which calls into each provider to do specific CUD (Create, Update, and Delete)
+operations on external resources.  Therefore, the managed reconciler is the
+ideal place to generate the change log entries very close to the source of
+truth.
+
+The managed reconciler calls the provider's [`ExternalClient.Observe()`
+method](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L914)
+before any CUD operation is performed, in order to understand the current state
+of the external system before the provider acts upon it. This up to date
+external state will be used to populate the "before" field in the change log
+entry.
+
+If a CUD operation is performed later on in the same `Reconcile()`, then we
+will perform an additional call to this same `Observe()` method directly after
+the CUD operation is performed. This second `Observe()` result will be used to
+populate the "after" field in the change log entry.
+
+Some helpful pointers to where CUD operations are performed in the managed
+reconciler:
+
+* [Create](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L1058)
+* [Update](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L1189)
+* [Delete](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L954)
+
+The code paths through the managed reconciler's `Reconcile()` method are already
+a bit complex. One design choice that has helped minimize the complexity and
+assist in readability is that the reconciler returns as early as it can after
+performing an operation. Since we will now be adding an additional `Observe()`
```

Comment:
Additional `Observe` means one more extra call to the cloud provider API and will increase probability of API ratelimiting. At least for `Create`s, we always ` queue` in the managed reconcilers and have a consecutive `Observe`, which we can leverage somehow.

Reply:
I definitely hear the concerns about the extra `Observe()` call. The usefulness of it is even being called into question with https://github.com/crossplane/crossplane/pull/5822#discussion_r1687250806 😊 

I'm not against cutting the "after" state completely, at least from this initial alpha implementation. I think we still tell an interesting story about WHY we made a change without having that data.

Not implementing an "after" state may be preferable instead of having to manage state across multiple reconciles to link together the `Observe()` results from a subsequent reconcile with the `Create()` or `Update()` operation from a previous reconcile. As things are currently implemented (https://github.com/crossplane/crossplane-runtime/compare/master...jbw976:crossplane-runtime:change-logs), everything is contained within a single reconcile and is definitely more simple for that 😇 

So in this alpha implementation, we could:

1. perform the extra `Observe()` call and see the feedback from early adopters
2. cut the "after" state functionality and do nothing
3. code something more complex that connects `Observe()` state to `Create()/Update()/Delete()` observations across reconciliations

Strong opinions from you all?

Reply:
I think alpha is a great time to see what the impact of the second `Observe()` call would be.  If it's bad we can always cut it when it goes beta.

Reply:
You could always put it under it's own nested flag so the feature could run with or without the second `Observe()` call.

Reply:
Chatted with @negz a bit more about "after" state and a 2nd `Observe()` call:

* From the first `Observe()` call, we should have a very up to date `status.atProvider` for any modern managed resource
* The difference between `spec.forProvider` and `status.atProvider` should fairly conclusively indicate why we are performing an operation
* The content of an "after" state collection from a 2nd `Observe()` is likely to not be any different from the "before", and therefore have limited value as called out in https://github.com/crossplane/crossplane/pull/5822#discussion_r1687250806

Therefore, it's reasonable to remove from this design the "after" state and thus any reason to perform a 2nd `Observe()`.  Without it, we still have a useful record of why we are making changes, whether that be because the spec's desired state has changed, we detected drift in the external resource, etc.

Perhaps Providers could be extended further in the future to provide more details in their [`ExternalObservation`](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L399) about why they concluded `ResourceUpToDate: false` and we can include that information in the change logs. For example, there is a `ExternalObservation.Diff` field, but I don't believe many Providers are currently setting it.